### PR TITLE
Minor tensorboard callback fixes

### PR DESCRIPTION
**Pull request by enkait**
1. The issue described in https://github.com/fchollet/keras/issues/7397
I.e. the embedding layer has gradients represented by an IndexedSlice as opposed to a tensor. The dense gradient tensor expected by tensorflow should be extracted with the .values function.
2. Fail loudly if validation_data is None and histogram_freq is non-zero.

### CODE REVIEW

**Comment by fchollet** [comment]
Please fix failing tests: https://travis-ci.org/fchollet/keras/builds/262394914

File: keras/callbacks.py [hunk 1]
```diff
@@ -741,6 +746,9 @@ def set_model(self, model):
     def on_epoch_end(self, epoch, logs=None):
         logs = logs or {}
 
+        if not self.validation_data and self.histogram_freq:
```
Comment (Dref360) [inline keras/callbacks.py]: Please add a test to validate that this exception is raised.

**Comment by FrugoFruit90** [comment]
I tried your suggestion using my toy example (https://github.com/fchollet/keras/issues/7489)
Not sure what is the reason, here is the traceback:
```
Traceback (most recent call last):
  File "/home/magda/Desktop/a.py", line 32, in <module>
    epochs=2, batch_size=64, validation_split=0.2, shuffle=True)
  File "/home/magda/envs/test/local/lib/python2.7/site-packages/keras/engine/training.py", line 1507, in fit
    initial_epoch=initial_epoch)
  File "/home/magda/envs/test/local/lib/python2.7/site-packages/keras/engine/training.py", line 1156, in _fit_loop
    outs = f(ins_batch)
  File "/home/magda/envs/test/local/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 2269, in __call__
    **self.session_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 789, in run
    run_metadata_ptr)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 997, in _run
    feed_dict_string, options, run_metadata)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 1132, in _do_run
    target_list, options, run_metadata)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 1152, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Input to reshape is a tensor with 64000 values, but the requested shape requires a multiple of 3000
	 [[Node: lstm_1/Reshape_7 = Reshape[T=DT_FLOAT, Tshape=DT_INT32, _device="/job:localhost/replica:0/task:0/cpu:0"](lstm_1/BiasAdd_3, lstm_1/stack_3)]]

Caused by op u'lstm_1/Reshape_7', defined at:
  File "/home/magda/Desktop/a.py", line 16, in <module>
    lstm_out = LSTM(10)(x)
  File "/home/magda/envs/test/local/lib/python2.7/site-packages/keras/layers/recurrent.py", line 262, in __call__
    return super(Recurrent, self).__call__(inputs, **kwargs)
  File "/home/magda/envs/test/local/lib/python2.7/site-packages/keras/engine/topology.py", line 596, in __call__
    output = self.call(inputs, **kwargs)
  File "/home/magda/envs/test/local/lib/python2.7/site-packages/keras/layers/recurrent.py", line 333, in call
    preprocessed_input = self.preprocess_input(inputs, training=None)
  File "/home/magda/envs/test/local/lib/python2.7/site-packages/keras/layers/recurrent.py", line 1086, in preprocess_input
    timesteps, training=training)
  File "/home/magda/envs/test/local/lib/python2.7/site-packages/keras/layers/recurrent.py", line 55, in _time_distributed_dense
    x = K.reshape(x, K.stack([-1, timesteps, output_dim]))
  File "/home/magda/envs/test/local/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 1734, in reshape
    return tf.reshape(x, shape)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/gen_array_ops.py", line 2451, in reshape
    name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 767, in apply_op
    op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 2506, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 1269, in __init__
    self._traceback = _extract_stack()

InvalidArgumentError (see above for traceback): Input to reshape is a tensor with 64000 values, but the requested shape requires a multiple of 3000
	 [[Node: lstm_1/Reshape_7 = Reshape[T=DT_FLOAT, Tshape=DT_INT32, _device="/job:localhost/replica:0/task:0/cpu:0"](lstm_1/BiasAdd_3, lstm_1/stack_3)]]
```

**Comment by enkait** [comment]
@fchollet, @Dref360: thank you both for reviewing!

@FrugoFruit90, I commented out the tensorboard callback in your change, and got the same error you described, which suggests to me that this change isn't the culprit.
Could you try to do the same?

**Comment by FrugoFruit90** [comment]
@enkait  I'm sorry, I didn't mean to suggest that your change broke anything! Just that unfortunately it didn't fix this particular problem, which seems very related :)
Keep up the good work!

File: keras/callbacks.py [hunk 2]
```diff
@@ -741,6 +747,9 @@ def set_model(self, model):
...
+            raise ValueError("If printing histograms, validation_data must be "
+                             "provided, and cannot be a generator.")
```
Comment (shadiakiki1986) [inline keras/callbacks.py]: Does this mean that tensorboard callback cannot be used with `fit_generator`?

**Comment by gokceneraslan** [comment]
@enkait isn't it just sufficient to put an [0] to get_gradients call like https://github.com/fchollet/keras/issues/7489#issuecomment-319380958?

File: keras/callbacks.py [hunk 3]
```diff
@@ -658,6 +658,12 @@ def set_model(self, model):
                     if self.write_grads:
                         grads = model.optimizer.get_gradients(model.total_loss,
                                                               weight)
+
+                        def is_indexed_slices(grad):
+                            return type(grad).__name__ == 'IndexedSlices'
+                        grads = [
+                            grad.values if is_indexed_slices(grad) else grad
+                            for grad in grads]
```
Comment (gokceneraslan) [inline keras/callbacks.py]: I don't think that's really necessary here. See https://github.com/fchollet/keras/issues/7489#issuecomment-319380958 for a simpler fix.

**Comment by enkait** [comment]
@shadiakiki1986 fit_generator can take validation_data as a tensor, so that shouldn't be an issue. Only if you pass a generator as validation_data will there be a problem.

@gokceneraslan Good question! I don't know enough about tensorflow or keras to be sure.
I was expecting from the [documentation](https://www.tensorflow.org/api_docs/python/tf/summary/histogram) that tf.summary.histogram has to have a normal Tensor, but your solution seems to work fine for the problem in #7397.
The [get_gradients](https://github.com/fchollet/keras/blob/master/keras/optimizers.py#L72) code is written assuming that there can be more than one element in the returned grads, but you may also be right that really there will only ever be one element.
So I guess send a pull request, and we'll see what fchollet says. :)

**Pull request merged**